### PR TITLE
p2p/discover: implement ENR node filtering

### PR DESCRIPTION
**Pull request by MatusKysel**
### Description

This PR implements [BEP194](https://github.com/bnb-chain/BEPs/pull/194)

### Rationale

This change should improve peer discovery time by filtering out peers from different networks e.g. ETH

### Changes

Notable changes: 
* adding peer do the table is now async as it also requires to get ENR record from the node
* this could also help to filter out nodes, that are not running latest HF

### CODE REVIEW

File: cmd/bootnode/main.go [hunk 1]
```diff
@@ -148,3 +161,31 @@ func printNotice(nodeKey *ecdsa.PublicKey, addr net.UDPAddr) {
...
+	var filter forkid.Filter
+	switch *chain {
+	case "bsc":
+		filter = forkid.NewStaticFilter(params.BSCChainConfig, params.BSCGenesisHash)
+	case "chapel":
+		filter = forkid.NewStaticFilter(params.ChapelChainConfig, params.ChapelGenesisHash)
+	case "rialto":
+		filter = forkid.NewStaticFilter(params.RialtoChainConfig, params.RialtoGenesisHash)
+	case "yolo":
+		filter = forkid.NewStaticFilter(params.YoloV3ChainConfig, params.YoloV3GenesisHash)
```
Comment (bnb-twu) [inline cmd/bootnode/main.go:175]: we have a network called YoloV3? 😂

File: cmd/bootnode/main.go [hunk 2]
```diff
@@ -148,3 +161,31 @@ func printNotice(nodeKey *ecdsa.PublicKey, addr net.UDPAddr) {
 	fmt.Println("Note: you're using cmd/bootnode, a developer tool.")
 	fmt.Println("We recommend using a regular node as bootstrap node for production deployments.")
 }
+
+func parseEthFilter(chain *string) (discover.NodeFilter, error) {
```
Comment (bnb-twu) [inline cmd/bootnode/main.go:165]: no unit tests for this function?

File: p2p/discover/table.go [hunk 3]
```diff
@@ -339,10 +342,16 @@ func (tab *Table) doRevalidate(done chan<- struct{}) {
 
 	// Also fetch record if the node replied and returned a higher sequence number.
 	if last.Seq() < remoteSeq {
-		n, err := tab.net.RequestENR(unwrapNode(last))
-		if err != nil {
+		n, enr_err := tab.net.RequestENR(unwrapNode(last))
+		if enr_err != nil {
 			tab.log.Debug("ENR request failed", "id", last.ID(), "addr", last.addr(), "err", err)
```
Comment (bnb-twu) [inline p2p/discover/table.go:347]:
```suggestion
			tab.log.Debug("ENR request failed", "id", last.ID(), "addr", last.addr(), "err", enr_err)
```
Don't we need to log `enr_err`?

File: p2p/discover/table.go [hunk 4]
```diff
@@ -339,10 +342,16 @@ func (tab *Table) doRevalidate(done chan<- struct{}) {
 
 	// Also fetch record if the node replied and returned a higher sequence number.
 	if last.Seq() < remoteSeq {
-		n, err := tab.net.RequestENR(unwrapNode(last))
-		if err != nil {
+		n, enr_err := tab.net.RequestENR(unwrapNode(last))
+		if enr_err != nil {
```
Comment (bnb-twu) [inline p2p/discover/table.go:346]: nit: 
```suggestion
		n, enrErr := tab.net.RequestENR(unwrapNode(last))
		if enrErr != nil {
```

File: p2p/discover/table.go [hunk 5]
```diff
@@ -502,6 +521,21 @@ func (tab *Table) addSeenNode(n *node) {
...
+		if err != nil {
+			tab.log.Debug("ENR request failed", "id", n.ID(), "addr", n.addr(), "err", err)
+		} else {
+			if !tab.enrFilter(node.Record()) {
+				tab.log.Trace("ENR record filter out", "id", n.ID(), "addr", n.addr())
+				return true
+			}
+		}
+	}
+	return false
```
Comment (bnb-twu) [inline p2p/discover/table.go:536]:
```suggestion
	if tab.enrFilter != nil {
		node, err := tab.net.RequestENR(unwrapNode(n))
		if err != nil {
			tab.log.Debug("ENR request failed", "id", n.ID(), "addr", n.addr(), "err", err)
			return false
		} else {
			if !tab.enrFilter(node.Record()) {
				tab.log.Trace("ENR record filter out", "id", n.ID(), "addr", n.addr())
				return true
			}
		}
	}
	return false
```

File: p2p/discover/table.go [hunk 6]
```diff
@@ -339,10 +342,16 @@ func (tab *Table) doRevalidate(done chan<- struct{}) {
...
-		if err != nil {
-			tab.log.Debug("ENR request failed", "id", last.ID(), "addr", last.addr(), "err", err)
+		n, enrErr := tab.net.RequestENR(unwrapNode(last))
+		if enrErr != nil {
+			tab.log.Debug("ENR request failed", "id", last.ID(), "addr", last.addr(), "err", enrErr)
 		} else {
+			if tab.enrFilter != nil {
+				if !tab.enrFilter(n.Record()) {
+					tab.log.Trace("ENR record filter out", "id", last.ID(), "addr", last.addr())
+					err = fmt.Errorf("filtered node")
```
Comment (Mister-EA) [inline p2p/discover/table.go:352]: Why is it an error if a node gets filtered out? Or was it not supposed to end up in the table in the first place?

File: p2p/discover/table_test.go [hunk 7]
```diff
@@ -382,7 +385,7 @@ func TestTable_revalidateSyncRecord(t *testing.T) {
 	r.Set(enr.IP(net.IP{127, 0, 0, 1}))
 	id := enode.ID{1}
 	n1 := wrapNode(enode.SignNull(&r, id))
-	tab.addSeenNode(n1)
+	tab.addSeenNodeSync(n1)
```
Comment (Mister-EA) [inline p2p/discover/table_test.go:388]: Ideally other tests should not be changed when testing new functionality. Is there a reason it was necessary to modify these tests?

File: p2p/discover/table.go [hunk 8]
```diff
@@ -502,6 +521,22 @@ func (tab *Table) addSeenNode(n *node) {
 	}
 }
 
+func (tab *Table) filterNode(n *node) bool {
+	if tab.enrFilter != nil {
+		node, err := tab.net.RequestENR(unwrapNode(n))
+		if err != nil {
+			tab.log.Debug("ENR request failed", "id", n.ID(), "addr", n.addr(), "err", err)
+			return false
+		} else {
```
Comment (emailtovamos) [inline p2p/discover/table.go:530]: Is the `else` necessary here?

File: p2p/discover/table.go [hunk 9]
```diff
@@ -502,6 +521,22 @@ func (tab *Table) addSeenNode(n *node) {
 	}
 }
 
+func (tab *Table) filterNode(n *node) bool {
+	if tab.enrFilter != nil {
```
Comment (alexlucaci) [inline p2p/discover/table.go:525]: I would invert the if condition following an early exit approach. i.e.: `if tab.enrFilter == nil { return false }` this way the logic block will not be indented anymore which increases code readability

File: p2p/discover/table_test.go [hunk 10]
```diff
@@ -396,6 +399,38 @@ func TestTable_revalidateSyncRecord(t *testing.T) {
...
+
+	enrFilter, _ := ParseEthFilter("bsc")
+
+	// Check test ENR record
+	var r1 enr.Record
+	r1.Set(enr.WithEntry("foo", "bar"))
+	if enrFilter(&r1) {
+		t.Fatalf("filterNode doesn't work correctly for entry")
+	}
+
```
Comment (alexlucaci) [inline p2p/discover/table_test.go:418]: Maybe we want to increase the testing logging verbosity. I would add an `t.Logf()` after each successful test step. This helps when reading a failed test.

File: p2p/discover/common.go [hunk 11]
```diff
@@ -35,18 +39,49 @@ type UDPConn interface {
 	LocalAddr() net.Addr
 }
 
+type NodeFilter func(*enr.Record) bool
```
Comment (alexlucaci) [inline p2p/discover/common.go:42]: As this type declares a function, I would rename it type to `NodeFilterFunc`

File: p2p/server.go [hunk 12]
```diff
@@ -601,13 +605,28 @@ func (srv *Server) setupDiscovery() error {
 			unhandled = make(chan discover.ReadPacket, 100)
 			sconn = &sharedUDPConn{conn, unhandled}
 		}
+		f := func(r *enr.Record) bool {
+			if srv.forkFilter != nil {
```
Comment (alexlucaci) [inline p2p/server.go:609]: Inverting the if condition here also helps with readability as the logic block will no be indented anymore

File: p2p/discover/table_test.go (hunk 7 above)
Comment (Mister-EA) [reply p2p/discover/table_test.go:388]: Are you still planning to switch back to `addSeenNode` ?

File: p2p/discover/v4_udp.go [hunk 13]
```diff
@@ -42,7 +42,7 @@ var (
 	errExpired          = errors.New("expired")
 	errUnsolicitedReply = errors.New("unsolicited reply")
 	errUnknownNode      = errors.New("unknown node")
-	errTimeout          = errors.New("RPC timeout")
+	errTimeout          = errors.New("udp timeout")
```
Comment (Mister-EA) [inline p2p/discover/v4_udp.go:45]: What is the reason for this change?

File: p2p/discover/v5_udp.go [hunk 14]
```diff
@@ -164,7 +164,7 @@ func newUDPv5(conn UDPConn, ln *enode.LocalNode, cfg Config) (*UDPv5, error) {
 		closeCtx:       closeCtx,
 		cancelCloseCtx: cancelCloseCtx,
 	}
-	tab, err := newTable(t, t.db, cfg.Bootnodes, cfg.Log)
+	tab, err := newTable(t, t.db, cfg.Bootnodes, cfg.Log, nil)
```
Comment (Mister-EA) [inline p2p/discover/v5_udp.go:167]: Why  in `v4_udp.go` `cfg.FilterFunction` is passed, whereas here `nil` is passed?

File: p2p/server.go [hunk 15]
```diff
@@ -601,13 +605,28 @@ func (srv *Server) setupDiscovery() error {
...
-			Unhandled:   unhandled,
-			Log:         srv.log,
+			PrivateKey:     srv.PrivateKey,
+			NetRestrict:    srv.NetRestrict,
+			Bootnodes:      srv.BootstrapNodes,
+			Unhandled:      unhandled,
+			Log:            srv.log,
+			FilterFunction: f,
 		}
+		// TODO: create here filter function based on nodes genesis
```
Comment (Mister-EA) [inline p2p/server.go:629]: Is this left as a TODO on purpose?

File: p2p/discover/v5_udp.go (hunk 14 above)
Comment (Mister-EA) [reply p2p/discover/v5_udp.go:167]: It makes sense to add it I think. By default it should be `nil` anyway.

**Review by Mister-EA — APPROVED** [review]

File: cmd/bootnode/main.go [hunk 16]
```diff
@@ -35,19 +35,21 @@ import (
...
+		genKey        = flag.String("genkey", "", "generate a node key")
+		writeAddr     = flag.Bool("writeaddress", false, "write out the node's public key and quit")
+		nodeKeyFile   = flag.String("nodekey", "", "private key filename")
+		nodeKeyHex    = flag.String("nodekeyhex", "", "private key as hex (for testing)")
+		natdesc       = flag.String("nat", "none", "port mapping mechanism (any|none|upnp|pmp|extip:<IP>)")
+		netrestrict   = flag.String("netrestrict", "", "restrict network communication to the given IP networks (CIDR masks)")
+		runv5         = flag.Bool("v5", false, "run a v5 topic discovery bootnode")
+		verbosity     = flag.Int("verbosity", int(log.LvlInfo), "log verbosity (0-5)")
+		vmodule       = flag.String("vmodule", "", "log verbosity pattern")
+		networkFilter = flag.String("network", "", "<bsc/chapel/rialto/yolo> filters nodes by eth ENR entry")
```
Comment (brilliant-lx) [inline cmd/bootnode/main.go:48]: no yolo, right?

**Review by brilliant-lx — APPROVED** [review]

**Review by Mister-EA — APPROVED** [review]
